numba/targets/arraymath.py

```diff
@@ -668,6 +668,83 @@ def _trivial_isnan(x):
         return _trivial_isnan
 
 
+@overload(np.iscomplex)
+def np_iscomplex(x):
+
+    if not type_can_asarray(x):
+        raise TypingError("First argument must be array-like")
+
+    dt = determine_dtype(x)
+    if np.issubdtype(dt, np.complexfloating):
+        def impl(x):
+            ax = np.asarray(x) # NumPy uses asanyarray here!
+            return ax.imag != 0
+    else:
+        def impl(x):
+            ax = np.asarray(x)
+            zeros = np.zeros(ax.shape, dtype=types.boolean)
+            return zeros[()] # hack when x is a scalar value
+    return impl
+
+
+@overload(np.isreal)
+def np_isreal(x):
+    def impl(x):
+        ax = np.iscomplex(x)
+        return np.invert(ax)
+    return impl
+
+
+@overload(np.iscomplexobj)
+def iscomplexobj(x):
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/lib/type_check.py#L282-L320
+    dt = determine_dtype(x)
+    if np.issubdtype(dt, np.complexfloating):
+        @register_jitable
+        def impl(x):
+            return True
+    else:
+        @register_jitable
+        def impl(x):
+            return False
+    return impl
+
+
+@overload(np.isrealobj)
+def isrealobj(x):
+    # Return True if x is a not complex type or an array of complex numbers.
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/ccfbcc1cd9a4035a467f2e982a565ab27de25b6b/numpy/lib/type_check.py#L290-L322
+    def impl(x):
+        return not np.iscomplexobj(x)
+    return impl
+
+
+@overload(np.isneginf)
+def isneginf(x, out=None):
+
+    if not type_can_asarray(x):
+        raise TypingError("First argument must be array-like")
+
+    if numpy_version <= (1, 12):
+        wrapper = register_jitable(lambda x: np.asarray(x))
+    else:
+        wrapper = register_jitable(lambda x: x[()])
+
+    if is_nonelike(out):
+        def impl(x, out=None):
+            x = np.asarray(x)
+            out = np.zeros(x.shape, dtype=types.boolean)
+            np.logical_and(np.isinf(x), np.signbit(x), out)
+            return wrapper(out)
```

Comment:
I'm not sure about this. Why are scalars being converted to arrays, then requiring a heap allocation (np.zeros) and then a call to make them scalar again? Perhaps this implementation ought to branch at typing time on whether the input is scalar? It seems like it'd be costly to use e.g. `isneginf` in a tight loop else as I'm not convinced optimisers would be able to figure out how to optimise this indirection away.

Reply:
I've pushed a minor change in the code to prevent a heap allocation when `x` is a scalar value.